symengine/symengine_rcp.h

```diff
@@ -198,6 +198,73 @@ using Teuchos::print_stack_on_segfault;
 #endif
 
 
+template <class T>
+class EnableRCPFromThis {
+// Public interface
+public:
+    //! Get RCP<T> pointer to self (it will cast the pointer to T)
+    inline RCP<T> rcp_from_this() {
+#if defined(WITH_SYMENGINE_RCP)
+        return rcp(static_cast<T*>(this));
+#else
+        return rcp_static_cast<T>(weak_self_ptr_.create_strong());
+#endif
+    }
+
+    //! Get RCP<const T> pointer to self (it will cast the pointer to const T)
+    inline RCP<const T> rcp_from_this() const {
```

Comment:
Ondrej,

You need the const and non-const functions to be const-correct.  That is, if a client as a const reference to the object, then it must give up `RCP<const T>`, not `RCP<T>` when it calls `rcp_from_this()`.  Otherwise, that would violate const correctness (which is very important in C++).   See Item 15 "Use const proactively" in "C++ Coding Standards".